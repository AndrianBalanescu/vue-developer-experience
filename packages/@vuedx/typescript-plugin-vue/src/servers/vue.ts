import { TS } from '../interfaces';
import { createTemplateLanguageServer } from './template';
import { VirtualTextDocument, isNumber, isVirtualFile, getContainingFile } from 'vue-virtual-textdocument/src';
import { CreateLanguageServiceOptions } from '../types';
import { noop } from './noop';

export function createVueLanguageServer(options: CreateLanguageServiceOptions): TS.LanguageService {
  const template = createTemplateLanguageServer(options);
  const { helpers: h, service: script, context } = options;
  function choose(document: VirtualTextDocument) {
    return h.isRenderFunctionDocument(document) ? template : script;
  }
  function getTextSpan(document: VirtualTextDocument, span: TS.TextSpan): TS.TextSpan {
    if (h.isRenderFunctionDocument(document)) {
      const result = document.getOriginalOffsetAt(span.start);
      if (result) return { start: result.offset, length: result.length };
    }

    return span;
  }

  return {
    ...noop,
    getRenameInfo(fileName, position, options) {
      const document = h.getDocumentAt(fileName, position);

      if (!document) {
        return {
          canRename: false,
          localizedErrorMessage: '',
        };
      }

      return choose(document).getRenameInfo(document.fsPath, position, options);
    },

    findRenameLocations(fileName, position, findInStrings, findInComments) {
      const document = h.getDocumentAt(fileName, position);
      if (!document) return;

      return choose(document)
        .findRenameLocations(document.fsPath, position, findInStrings, findInComments)
        ?.map((item) => {
          options.context.log('xxx.findRenameLocations ' + JSON.stringify(item));

          if (isVirtualFile(item.fileName)) {
            item.originalContextSpan = item.contextSpan;
            item.originalTextSpan = item.textSpan;
            item.originalFileName = item.fileName;

            const virtual = h.getDocument(item.fileName) as VirtualTextDocument;

            item.fileName = virtual.container.fsPath;
            item.textSpan = getTextSpan(virtual, item.textSpan);
            if (item.contextSpan) item.contextSpan = getTextSpan(virtual, item.contextSpan);
          }

          options.context.log('yyy.findRenameLocations ' + JSON.stringify(item));

          return item;
        });
    },

    getApplicableRefactors(fileName, positionOrRange, preferences) {
      const document = h.getDocumentAt(fileName, isNumber(positionOrRange) ? positionOrRange : positionOrRange.pos);
      if (document) {
        return choose(document).getApplicableRefactors(document.fsPath, positionOrRange, preferences);
      }

      return [];
    },
  };
}
