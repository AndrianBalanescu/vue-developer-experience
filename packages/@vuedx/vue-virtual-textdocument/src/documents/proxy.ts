import { TextDocument, Position, Range } from 'vscode-languageserver-textdocument';
import { asFsPath, isNumber } from '../utils';

export abstract class ProxyTextDocument implements TextDocument {
  public readonly fsPath: string;

  public constructor(protected doc: TextDocument) {
    this.fsPath = asFsPath(this.doc.uri);
  }

  public get version() {
    this.refresh();

    return this.doc.version;
  }

  public get languageId() {
    this.refresh();

    return this.doc.languageId;
  }

  public get lineCount() {
    this.refresh();

    return this.doc.lineCount;
  }

  public get uri() {
    this.refresh();

    return this.doc.uri;
  }

  public getText(): string;
  public getText(range: Range): string;
  public getText(offset: number, length: number): string;
  public getText(range?: number | Range, length?: number): string {
    this.refresh();

    if (isNumber(range)) {
      return this.doc.getText().substr(range, length);
    } else {
      return this.doc.getText(range);
    }
  }

  public positionAt(offset: number): Position {
    this.refresh();

    return this.doc.positionAt(offset);
  }

  public offsetAt(position: Position): number {
    this.refresh();

    return this.doc.offsetAt(position);
  }

  protected refresh(): void {}
}
