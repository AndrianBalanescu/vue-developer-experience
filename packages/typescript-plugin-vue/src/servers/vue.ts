import { isNumber, isVirtualFile, isVueFile, VirtualTextDocument } from '@vuedx/vue-virtual-textdocument';
import { PluginConfig, TS } from '../interfaces';
import { LanguageServiceOptions } from '../types';
import { noop } from './noop';
import { createTemplateLanguageServer } from './template';

type GetElementType<T> = T extends (infer U)[] ? U : T;
export function createVueLanguageServer(options: LanguageServiceOptions): TS.LanguageService {
  const template = createTemplateLanguageServer(options);
  const { helpers: h, service: script, context } = options;

  function isFeatureEnabled<K extends keyof PluginConfig['features']>(
    featureName: K,
    checkFor: boolean | GetElementType<PluginConfig['features'][K]> = true
  ): boolean {
    const feature = context.config.features[featureName];

    return Array.isArray(feature) ? feature.includes(checkFor) : feature === checkFor;
  }

  function choose(document: VirtualTextDocument) {
    return h.isRenderFunctionDocument(document) ? template : script;
  }

  return {
    ...noop,

    organizeImports(scope, formatOptions, preferences) {
      if (!isFeatureEnabled('organizeImports')) return [];

      const document = h.getVueDocument(scope.fileName);
      if (document) {
        const virtual = document.getDocument('scriptSetup') || document.getDocument('script');
        if (virtual) {
          return script.organizeImports({ ...scope, fileName: virtual.fsPath }, formatOptions, preferences);
        }
      }

      return [];
    },

    getSemanticDiagnostics(fileName) {
      if (!isFeatureEnabled('diagnostics', 'semantic')) return [];

      const document = h.getVueDocument(fileName);
      const diagnostics: TS.Diagnostic[] = [];
      if (document) {
        ['script', '_render'].forEach((selector) => {
          const virtual = document.getDocument(selector);

          if (virtual) {
            const results = choose(virtual).getSemanticDiagnostics(virtual.fsPath);
            diagnostics.push(...results);
          }
        });
      }

      return diagnostics;
    },

    getQuickInfoAtPosition(fileName, position) {
      if (!isFeatureEnabled('quickInfo')) return;

      // TODO: Provide better quick info for components and props.
      const document = h.getDocumentAt(fileName, position);
      if (document) return choose(document).getQuickInfoAtPosition(document.fsPath, position);
    },

    getSuggestionDiagnostics(fileName) {
      if (!isFeatureEnabled('diagnostics', 'suggestion')) return [];

      const document = h.getVueDocument(fileName);

      const diagnostics: TS.DiagnosticWithLocation[] = [];
      if (document) {
        ['script', '_render'].forEach((selector) => {
          const virtual = document.getDocument(selector);

          if (virtual) diagnostics.push(...choose(virtual).getSuggestionDiagnostics(virtual.fsPath));
        });
      }

      return diagnostics;
    },

    getSyntacticDiagnostics(fileName) {
      if (!isFeatureEnabled('diagnostics', 'syntactic')) return [];
      const document = h.getVueDocument(fileName);

      const diagnostics: TS.DiagnosticWithLocation[] = [];
      if (document) {
        ['script', '_render'].forEach((selector) => {
          const virtual = document.getDocument(selector);

          if (virtual) diagnostics.push(...choose(virtual).getSyntacticDiagnostics(virtual.fsPath));
        });
      }

      return diagnostics;
    },

    getRenameInfo(fileName, position, options) {
      if (!isFeatureEnabled('rename')) return { canRename: false, localizedErrorMessage: 'Rename feature disabled.' };

      const document = h.getDocumentAt(fileName, position);

      if (!document) {
        return {
          canRename: false,
          localizedErrorMessage: 'Cannot find this Vue file.',
        };
      }

      return choose(document).getRenameInfo(document.fsPath, position, options);
    },

    findRenameLocations(fileName, position, findInStrings, findInComments) {
      if (!isFeatureEnabled('rename')) return [];

      const document = h.getVueDocument(fileName);
      if (!document) return;
      const block = document.blockAt(position);
      if (!block) return;

      const result: TS.RenameLocation[] = [];
      if (block.type === 'template') {
        const fromTemplate = template.findRenameLocations(
          document.getDocumentFileName('_render')!,
          position,
          findInStrings,
          findInComments
        );
        if (fromTemplate) result.push(...fromTemplate);
      }

      if (block.type === 'script') {
        const fromScript = script.findRenameLocations(
          document.getDocumentFileName('script')!,
          position,
          findInStrings,
          findInComments
        );
        if (fromScript) result.push(...fromScript);
      }

      result.sort((a, b) => a.textSpan.start - b.textSpan.start);

      return result;
    },

    getEditsForFileRename(oldFilePath, newFilePath, formatOptions, preferences) {
      if (!isFeatureEnabled('rename')) return [];

      console.log('FileRenamed > ' + oldFilePath);
      const document = h.getVueDocument(oldFilePath);
      const fileTextChanges: TS.FileTextChanges[] = [];
      const visited = new Set<string>();

      if (document) {
        const component = document.getDocument('_module');
        const currentChanges = script.getEditsForFileRename(
          component.fsPath,
          component.fsPath.replace(oldFilePath, newFilePath),
          formatOptions,
          preferences
        );
        console.log('Files Affected By Rename > ' + JSON.stringify(currentChanges.map(item => item.fileName)));

        fileTextChanges.push(...currentChanges);

        currentChanges.forEach((item) => {
          if (isVirtualFile(item.fileName) || isVueFile(item.fileName)) {
            const render = h.getVueDocument(item.fileName)?.getDocument('_render');
            if (render && !visited.has(render.fsPath)) {
              visited.add(render.fsPath);
              fileTextChanges.push(...template.getEditsForFileRenameIn(render.fsPath, oldFilePath, newFilePath));
            }
          }
        });
      }

      return fileTextChanges;
    },

    getApplicableRefactors(fileName, positionOrRange, preferences) {
      if (!isFeatureEnabled('refactor')) return [];

      const document = h.getDocumentAt(fileName, isNumber(positionOrRange) ? positionOrRange : positionOrRange.pos);
      const document2 = h.getDocumentAt(fileName, isNumber(positionOrRange) ? positionOrRange : positionOrRange.end);

      if (document && document === document2) {
        return choose(document).getApplicableRefactors(document.fsPath, positionOrRange, preferences);
      }

      return [];
    },

    getEditsForRefactor(fileName, formatOptions, positionOrRange, refactorName, actionName, preferences) {
      if (!isFeatureEnabled('refactor')) return;

      const document = h.getDocumentAt(fileName, isNumber(positionOrRange) ? positionOrRange : positionOrRange.pos);
      const document2 = h.getDocumentAt(fileName, isNumber(positionOrRange) ? positionOrRange : positionOrRange.end);

      if (document && document === document2) {
        return choose(document).getEditsForRefactor(
          document.fsPath,
          formatOptions,
          positionOrRange,
          refactorName,
          actionName,
          preferences
        );
      }
    },
  };
}
