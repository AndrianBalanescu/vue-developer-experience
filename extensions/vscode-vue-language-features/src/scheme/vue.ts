import vscode from 'vscode'
import { injectable } from 'inversify'
import { Installable } from '../utils/installable'
import { DocumentService } from '../services/documents'

@injectable()
export class VueVirtualDocumentProvider
  extends Installable
  implements vscode.TextDocumentContentProvider {
  constructor(private readonly documents: DocumentService) {
    super()
  }

  public install() {
    super.install()

    return vscode.Disposable.from(
      vscode.workspace.registerTextDocumentContentProvider('vue', this),
      this.documents.onDidChangeTextDocument(({ uri }) => {
        this.onDidChangeEmitter.fire(uri)
      }),
    )
  }

  private onDidChangeEmitter = new vscode.EventEmitter<vscode.Uri>()
  public onDidChange = this.onDidChangeEmitter.event

  async provideTextDocumentContent(request: vscode.Uri) {
    let uri = request.toString()

    if (uri.startsWith('vue:/') && !uri.startsWith('vue://')) {
      uri = uri.replace(/^vue:/, 'vue://')
    }

    try {
      const document = await this.documents.getVirtualDocument(uri)

      return document?.getText()
    } catch (error) {
      return `/*\nError: ${error.message}\n${error.stack}\n*/`
    }
  }
}
