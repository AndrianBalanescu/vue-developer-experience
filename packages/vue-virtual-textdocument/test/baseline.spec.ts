import { sync as glob } from 'fast-glob';
import Path from 'path';
import FS from 'fs';
import { VueTextDocument } from '../src/documents/vue';

expect.addSnapshotSerializer({
  test: (value) => typeof value === 'string',
  print: (value) => `${value}`,
});

describe('VueVirtualDocument/baseline', () => {
  const dir = Path.resolve(__dirname, '../../typescript-plugin-vue/tests/e2e/project/src');
  const fileNames = glob('**/*.vue', {
    cwd: dir,
    absolute: false,
  });

  test.each(fileNames)('%s', async (name) => {
    const fileName = Path.resolve(dir, name);
    const content = await FS.promises.readFile(fileName, { encoding: 'utf-8' });
    const document = VueTextDocument.create(`file://${fileName}`, 'vue', 0, content);

    expect(document.getDocument('_render')?.getText()).toMatchSnapshot();
  });
});
