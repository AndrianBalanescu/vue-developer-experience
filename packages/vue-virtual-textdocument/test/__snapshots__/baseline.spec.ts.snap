// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VueVirtualDocument/baseline javascript/src/App.vue 1`] = `
import _Ctx from './App.vue?internal'
import HelloWorld from './components/HelloWorld.vue'


export function render(_ctx: InstanceType<typeof _Ctx>) {
  return /*@@vue:start*/<main><h1>A sample vue application</h1><HelloWorld name="Rahul"/></main>/*@@vue:end*/
}
// Version 0

`;

exports[`VueVirtualDocument/baseline javascript/src/components/HelloWorld.vue 1`] = `
import _Ctx from './HelloWorld.vue?internal'


export function render({name}: InstanceType<typeof _Ctx>) {
  return /*@@vue:start*/<p>Hello {name}!</p>/*@@vue:end*/
}
// Version 0

`;

exports[`VueVirtualDocument/baseline javascript-with-config/src/App.vue 1`] = `
import _Ctx from './App.vue?internal'
import HelloWorld from './components/HelloWorld.vue'


export function render(_ctx: InstanceType<typeof _Ctx>) {
  return /*@@vue:start*/<main><h1>A sample vue application</h1><HelloWorld name="Rahul"/></main>/*@@vue:end*/
}
// Version 0

`;

exports[`VueVirtualDocument/baseline javascript-with-config/src/components/HelloWorld.vue 1`] = `
import _Ctx from './HelloWorld.vue?internal'


export function render({name}: InstanceType<typeof _Ctx>) {
  return /*@@vue:start*/<p>Hello {name}!</p>/*@@vue:end*/
}
// Version 0

`;

exports[`VueVirtualDocument/baseline typescript/src/App.vue 1`] = `
import _Ctx from './App.vue?internal'
import HelloWorld from './components/HelloWorld.vue'


export function render(_ctx: InstanceType<typeof _Ctx>) {
  return /*@@vue:start*/<main><h1>A sample vue application</h1><HelloWorld name="Rahul"/></main>/*@@vue:end*/
}
// Version 0

`;

exports[`VueVirtualDocument/baseline typescript/src/components/HelloWorld.vue 1`] = `
import _Ctx from './HelloWorld.vue?internal'


export function render({name}: InstanceType<typeof _Ctx>) {
  return /*@@vue:start*/<p>Hello {name}!</p>/*@@vue:end*/
}
// Version 0

`;

exports[`VueVirtualDocument/baseline typescript-with-config/src/App.vue 1`] = `
import _Ctx from './App.vue?internal'
import HelloWorld from './components/HelloWorld.vue'


export function render(_ctx: InstanceType<typeof _Ctx>) {
  return /*@@vue:start*/<main><h1>A sample vue application</h1><HelloWorld name="Rahul"/></main>/*@@vue:end*/
}
// Version 0

`;

exports[`VueVirtualDocument/baseline typescript-with-config/src/components/HelloWorld.vue 1`] = `
import _Ctx from './HelloWorld.vue?internal'


export function render({name}: InstanceType<typeof _Ctx>) {
  return /*@@vue:start*/<p>Hello {name}!</p>/*@@vue:end*/
}
// Version 0

`;
